### PR TITLE
[libc] fix issues around stack protector

### LINKED ISSUE #74487: [libc] hermetic build fails at `_start`
On my machine, `hermetic` tests are still failing from time to time. In debug build, such tests are failing right at `_start`.

`llvm-project/build/projects/libc/test/src/__support/libc.test.src.__support.endian_test.__hermetic__.__build__`.

Compiler:
```
clang version 16.0.6 (https://gitlab.archlinux.org/archlinux/packaging/packages/clang.git ff762c61c53dca880df8ed701c1bf9c73117e675)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

It seems that the _start function is attempting to read data based at `fs`, and that instruction triggers a SEGV.

image

### PR DESCRIPTION
**Pull request by SchrodingerZhu**
If a function is declared with stack-protector, the compiler may try to load the TLS. However, inside certain runtime functions, TLS may not be available. This patch disables stack protectors for such routines to fix the problem.

Closes #74487.

### CODE REVIEW

File: libc/startup/linux/aarch64/start.cpp [hunk 1]
```diff
@@ -193,6 +193,8 @@ __attribute__((noinline)) static void do_start() {
   LIBC_NAMESPACE::main_thread_attrib.atexit_callback_mgr =
       LIBC_NAMESPACE::internal::get_thread_atexit_callback_mgr();
 
+  LIBC_NAMESPACE::atexit(
```
Comment (michaelrj-google) [inline libc/startup/linux/aarch64/start.cpp:196]: please add a comment here (and in the other startup code) mentioning that this `atexit` call needs to be the first one.

File: libc/startup/linux/aarch64/start.cpp [hunk 2]
```diff
@@ -184,7 +184,7 @@ __attribute__((noinline)) static void do_start() {
     app.tls.align = phdr->p_align;
   }
 
-  LIBC_NAMESPACE::TLSDescriptor tls;
+  static LIBC_NAMESPACE::TLSDescriptor tls;
```
Comment (michaelrj-google) [inline libc/startup/linux/aarch64/start.cpp:187]: is changing this to static necessary?

File: libc/startup/linux/x86_64/CMakeLists.txt [hunk 3]
```diff
@@ -15,6 +15,7 @@ add_startup_object(
     libc.src.string.memory_utils.inline_memcpy
     libc.src.unistd.environ
   COMPILE_OPTIONS
+    -fno-stack-protector
```
Comment (michaelrj-google) [inline libc/startup/linux/x86_64/CMakeLists.txt:18]: should this also be added to the riscv startup options?

File: libc/src/__support/OSUtil/linux/quick_exit.h [hunk 4]
```diff
@@ -9,15 +9,17 @@
...
 #include "src/__support/common.h"
 
 #include <sys/syscall.h> // For syscall numbers.
 
 namespace LIBC_NAMESPACE {
 
-LIBC_INLINE void quick_exit(int status) {
+// mark as no_stack_protector since TLS can be torn down before calling
+// quick_exit
+__attribute__((no_stack_protector)) LIBC_INLINE void quick_exit(int status) {
```
Comment (nickdesaulniers) [inline libc/src/__support/OSUtil/linux/quick_exit.h:22]: should we use `[[gnu::no_stack_protector]]` style?

**Comment by nickdesaulniers** [comment]
I thought certain platforms have a global vs thread local stack canary?

File: libc/startup/linux/aarch64/start.cpp (hunk 2 above)
Comment (SchrodingerZhu) [reply libc/startup/linux/aarch64/start.cpp:187]: Unfortunately yes, the lambda expression to the `atexit` routine must be compatible with `void (*)()` hence capturing nothing.

File: libc/startup/linux/x86_64/CMakeLists.txt (hunk 3 above)
Comment (SchrodingerZhu) [reply libc/startup/linux/x86_64/CMakeLists.txt:18]: Waiting for specifications from @nickdesaulniers

**Comment by SchrodingerZhu** [comment]
@nickdesaulniers it does not seem to be arch dependent on linux: https://github.com/bminor/musl/blob/f314e133929b6379eccc632bef32eaebb66a7335/src/env/__stack_chk_fail.c#L21C6-L21C6

**Comment by SchrodingerZhu** [comment]
I updated the patch to disable stack protector explicitly for `x86-64`. The `atexit` changes are still applied to all archs since 1) it is more consistent 2) it may still be possible for other callbacks to use TLS for purposes other than stack protector.

File: libc/src/__support/OSUtil/linux/quick_exit.h (hunk 4 above)
Comment (SchrodingerZhu) [reply libc/src/__support/OSUtil/linux/quick_exit.h]: I tried just now: it seems that `[[clang::no_stack_protector]]` works but `[[gnu::no_stack_protector]]` is not recognized by `clang`.

**Comment by nickdesaulniers** [comment]
> > I thought certain platforms have a global vs thread local stack canary?
> 
> Could you please give a link to the specifications? I wish I can add links in the comment and gate the attributes accordingly.

This is only from observation; my prior work on the Linux kernel involved getting additional functionality added to clang for the kernel's more exotic stack protector implementation.  The kernel doesn't share the same ABI as userspace, so it can break its own intern ABI whenever, which allows for it to do better than the ABI constrained userspace.

I'd guess maybe the Itanium x86 psABI, ARM ARM, and RISCV psABI might have specific details, but I don't know.  I've only observed behavior by playing with the compiler.

File: libc/src/__support/OSUtil/linux/quick_exit.h [hunk 5]
```diff
@@ -9,15 +9,22 @@
...
 #include <sys/syscall.h> // For syscall numbers.
 
 namespace LIBC_NAMESPACE {
 
-LIBC_INLINE void quick_exit(int status) {
+// mark as no_stack_protector for x86 since TLS can be torn down before calling
+// quick_exit so that the stack protector canary cannot be loaded.
+#if defined(__i386) || defined(__i386__) || defined(__x86_64__) ||             \
+    defined(__x86_64)
+__attribute__((no_stack_protector))
```
Comment (nickdesaulniers) [inline libc/src/__support/OSUtil/linux/quick_exit.h:24]: can we use `[[]]` style attributes rather than the older extension?

Comment (SchrodingerZhu) [reply libc/src/__support/OSUtil/linux/quick_exit.h:24]: Sorry I have folded my initial reply.

in-tree clang does not seem to recognize [[gnu::no_stack_protector]].

[[clang::]] works but I will need to add other macro guards.

Comment (nickdesaulniers) [reply libc/src/__support/OSUtil/linux/quick_exit.h:24]: worth filing a bug against clang then. Clang implemented the attribute way before GNU did.

**Review by nickdesaulniers — APPROVED** [review]

File: libc/src/__support/OSUtil/linux/quick_exit.h (hunk 5 above)
Comment (SchrodingerZhu) [reply libc/src/__support/OSUtil/linux/quick_exit.h:24]: Issue submitted at https://github.com/llvm/llvm-project/issues/75235.

**Pull request merged**